build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy

```diff
@@ -42,20 +42,21 @@ class BinaryCompatibilityHelper {
         String currentVersion
     ) {
         japicmpTask.tap {
+            doNotTrackState("classloading issues with rules")
 
             addExcludeFilter(AnonymousClassesFilter)
             addExcludeFilter(KotlinInternalFilter)
 
             def acceptedChangesMap = acceptedViolations.toAcceptedChangesMap()
 
-            richReport.tap {
+            richReport.get().tap {
                 addRule(IncubatingInternalInterfaceAddedRule, [
                     acceptedApiChanges: acceptedChangesMap,
-                    publicApiPatterns: richReport.includedClasses
+                    publicApiPatterns: includedClasses.get()
                 ])
                 addRule(MethodsRemovedInInternalSuperClassRule, [
                     acceptedApiChanges: acceptedChangesMap,
-                    publicApiPatterns: richReport.includedClasses
+                    publicApiPatterns: includedClasses.get()
```

Comment:
Lots of eager resolution here; is that because the rule parameters like `publicApiPatterns` cannot accept a Provider?

Can we refactor `AbstractSuperClassChangesRule` like so? I'm not sure how we'd inject the ObjectFactory though.

```
-    private final List<Pattern> publicApiPatterns
+    private final ListProperty<Pattern> publicApiPatterns = objects.listProperty(Pattern)

```

Reply:
+1

Reply:
Hmm, technically we're only supposed to be passing a `String` here, so even this is already suspicious (thanks Groovy for not checking generics...). I could skip using a `Property` and just use a `Provider`, since this isn't really assignable.